src/appleseed/renderer/modeling/camera/fisheyelenscamera.cpp

```diff
@@ -0,0 +1,395 @@
+
+            // Compute ray origin and direction.
+            ray.m_org = transform.get_local_to_parent().extract_translation();
+            ray.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(ndc.get_value())));
+
+            // Compute ray derivatives.
+            if (ndc.has_derivatives())
+            {
+                const Vector2d px(ndc.get_value() + ndc.get_dx());
+                const Vector2d py(ndc.get_value() + ndc.get_dy());
+                ray.m_rx.m_org = ray.m_org;
+                ray.m_ry.m_org = ray.m_org;
+                ray.m_rx.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(px)));
+                ray.m_ry.m_dir = normalize(transform.vector_to_parent(-ndc_to_camera(py)));
+                ray.m_has_differentials = true;
+            }
+        }
+
+        bool connect_vertex(
+            SamplingContext&        sampling_context,
+            const float             time,
+            const Vector3d&         point,
+            Vector2d&               ndc,
+            Vector3d&               outgoing,
+            float&                  importance) const override
+        {
+            // Project the point onto the film plane.
+            if (!project_point(time, point, ndc))
+                return false;
+
+            // The connection is impossible if the projected point lies outside the film.
+            if (ndc[0] < 0.0 || ndc[0] >= 1.0 ||
+                ndc[1] < 0.0 || ndc[1] >= 1.0)
+                return false;
+
+            // Retrieve the camera transform.
+            Transformd scratch;
+            const Transformd& transform = m_transform_sequence.evaluate(time, scratch);
+
+            // Compute the outgoing direction vector in world space.
+            outgoing = point - transform.get_local_to_parent().extract_translation();
+
+            // Compute the emitted importance.
+            const Vector3d film_point = ndc_to_camera(ndc);
+            const double square_dist_film_lens = square_norm(film_point);
+            const double dist_film_lens = sqrt(square_dist_film_lens);
+            const double cos_theta = m_focal_length / dist_film_lens;
+            const double solid_angle = m_pixel_area * cos_theta / square_dist_film_lens;
+            importance = 1.0f / static_cast<float>(square_norm(outgoing) * solid_angle);
+
+            // The connection was possible.
+            return true;
+        }
+
+      private:
+        enum Projection
+        {
+            EquisolidAngle, 
+            Equidistant, 
+            Stereographic, 
+            Thoby
+        };
+
+        Projection m_projection_type;
+
+        Vector3d ndc_to_camera(const Vector2d& point) const
+        {
+            const double x = (0.5 - point.x) * m_film_dimensions[0];
+            const double y = (point.y - 0.5) * m_film_dimensions[1];
+
+            const double radius_1 = sqrt(x * x + y * y);
+            const double rcp_radius_1 = 1 / radius_1;
```

Comment:
`1` -> `1.0` (same thing for all constants that appear in floating point expressions).